Add standard TimeTool

kpavlov opened the pull request:
# Add standard TimeTool

- Implement TimeTool + unit test
- Update QuickStart guide and [`Tool` KDoc](https://github.com/JetBrains/koog/pull/387/files#diff-88141b61c9a662860a93872a719104fa58a5338b666c7530c58b60c60f5a3578R59)

#### Type of the change
- [x] New feature
- [ ] Bug fix
- [x] Documentation fix

#### Checklist for all pull requests
- [x] The pull request has a description of the proposed change
- [x] I read the [Contributing Guidelines](https://github.com/JetBrains/koog/blob/main/CONTRIBUTING.md) before opening the pull request
- [x] The pull request uses **`develop`** as the base branch
- [x] Tests for the changes have been added
- [ ] All new and existing tests passed

##### Additional steps for pull requests adding a new feature
- [ ] An issue describing the proposed change exists
- [ ] The pull request includes a link to the issue
- [ ] The change was discussed and approved in the issue
- [x] Docs have been added / updated

kpavlov committed:
Add standard TimeTool

- Implement TimeTool
- Test it with [kotest-assertions](https://kotest.io/docs/assertions/assertions.html)
- Update QuickstartGuide.md

kpavlov commented on agents/agents-ext/src/commonMain/kotlin/ai/koog/agents/ext/tool/TimeTool.kt:
```diff
@@ -0,0 +1,93 @@
+ * Defaults to the system's default timezone.
+ * @param truncateToSeconds A flag to determine whether the time should be truncated to seconds. Defaults to `true`.
+ */
+public class TimeTool(
```
Should this class somehow make use of the [TimeProvider](https://github.com/JetBrains/koog/blob/develop/agents/agents-features/agents-features-memory/src/commonMain/kotlin/ai/koog/agents/memory/model/TimeProvider.kt) 🤔

sdubov commented on agents/agents-ext/src/commonMain/kotlin/ai/koog/agents/ext/tool/TimeTool.kt:
```diff
@@ -0,0 +1,93 @@
+    override val argsSerializer: KSerializer<Args> = Args.serializer()
+
+    override val descriptor: ToolDescriptor = ToolDescriptor(
+        name = "__time__",
```
Could you please clarify why do you use a special formatting for a tool name with "__" as a prefix and postfix?

sdubov commented on agents/agents-ext/src/commonTest/kotlin/ai/koog/agents/ext/tool/TimeToolTest.kt:
```diff
@@ -0,0 +1,103 @@
+            currentTime
+        }.toLocalDateTime(timezone)
+
+        println(toolResponse)
```
Maybe it worth adding it to the assertion custom message, or relay on the standard test framework error message? If there is some special case of this `println`, I would suggest to define a logger and write this message to a logger instead of stdout:

kpavlov replied:
Agree

sdubov commented on agents/agents-ext/src/commonTest/kotlin/ai/koog/agents/ext/tool/TimeToolTest.kt:
```diff
@@ -0,0 +1,103 @@
+
+    @Test
+    fun testTimeToolWithSpecificTimezone() = runTest {
+        for (timezone in listOf("UTC", "Europe/Paris", "+01:00")) {
```
It seems for me we should split this to a different tests to verify and get separate errors about each failure. Wdyt?

kpavlov replied:
I agree. I'd also prefer to convert it to `@ParametrizedTest`. Do you know something suitable for multiplatform tests. I can't use JUnit Params in common module. Otherwise, it might not make much sense splitting it into different test methods: good error message should be enough to diagnose.

kpavlov commented on gradle/libs.versions.toml:
```diff
@@ -50,6 +51,7 @@ mcp = { module = "io.modelcontextprotocol:kotlin-sdk", version.ref = "mcp" }
 slf4j-simple = { module = "org.slf4j:slf4j-simple", version.ref = "slf4j" }
 jetsign-gradle-plugin = { module = "com.jetbrains:jet-sign", version.ref = "jetsign" }
 testcontainers = { module = "org.testcontainers:testcontainers", version.ref = "testcontainers" }
+kotest-assertions-core = { module = "io.kotest:kotest-assertions-core", version.ref = "kotest" }
```
Thanks for the feedback, makes sense to aim for consistency.

That said, I proposed [Kotest assertions](https://kotest.io/docs/assertions/assertions.html) because they offer a more fluent, expressive API compared to kotlin.test and JUnit. They improve readability, reduce boilerplate, and help avoid common pitfalls like parameter ambiguity in Junit/KotlinTest `assertEquals(expected, actual)`—which is especially error-prone and hard to catch.

In Java, AssertJ is often used solve this. Kotest brings similar clarity and safety to Kotlin.

Happy to align with the team direction, but I believe Kotest adds real value, especially for larger test suites.
Would you be open to review Kotest APIs and consider it for wider adoption in the project?

kpavlov replied:
@aozherelyeva , let me [break one test](https://github.com/JetBrains/koog/pull/387/commits/8245a4c3fb8c2c9d9e25ba6a2faa266379184c7b) so you can see. I also added [soft asserts](https://kotest.io/docs/assertions/soft-assertions.html) for demonstration

aozherelyeva replied:
Well, this output indeed looks pretty good. 

@sdubov we had a discussion with @kpavlov, and he suggested trying to use `kotest-assertions` in other modules as well – as the output indeed looks more readable with them. Let's give it a try? If it works for us, I'll start updating the code accordingly. And if it doesn't, updating assertions here would be pretty easy. WDYT?

sdubov replied:
@kpavlov, thank you for your detailed examples. I got your comments about readability. 

I have several notes from my side related to this topic:

1. Any new dependency that we bring to the SDK solves some particular problem. As far as I know, we did not notice any issues with current test approach that we have. @aozherelyeva , @Ololoshechkin , please correct me if you have any issues with `kotlin.test` assertions on your side;
2. Adding new dependency introduce extra load on release process: you need to care about versioning, need to perform security checks, etc.;
3. From my perspective, it is better to have consistency across all tests (using one assertion framework) rather than having a test class with inconsistent, but more readable assertions. It might be quite notifiable, it you start using different libraries and on each new test class have no clear understanding what should I use;
4. If we decide to go with a new assertion library, I would prefer doing it as a separate PR, with (not necessarily, but I would vote for) updating existing test code for consistency.

kpavlov committed:
Merge branch 'develop' into time-tool

kpavlov committed:
Enhance TimeTool tests with `withClue` for improved error diagnostics

remove println

kpavlov committed:
Intentionally breaking tests

kpavlov committed:
Fix test

kpavlov committed:
Merge branch 'JetBrains:develop' into time-tool

kpavlov commented on gradle/libs.versions.toml:
```diff
@@ -92,4 +94,4 @@ kotlin-serialization = { id = "org.jetbrains.kotlin.plugin.serialization", versi
 kotlin-spring = { id = "org.jetbrains.kotlin.plugin.spring", version.ref = "kotlin" }
-spring-management = { id = "io.spring.dependency-management", version.ref = "spring-management" }
\ No newline at end of file
+spring-management = { id = "io.spring.dependency-management", version.ref = "spring-management" }
```
```suggestion
```

kpavlov committed:
Apply suggestion from @kpavlov

kpavlov committed:
Update libs.versions.toml

kpavlov committed:
Merge branch 'JetBrains:develop' into time-tool

kpavlov committed:
Merge branch 'develop' into time-tool

kpavlov committed:
Address PR comments

- Removed kotest assertions and dependency
- Changed TimeTool name
- Refactored TimeToolTest

kpavlov committed:
Updated Tool KDoc
